fix: `DurationParser` allowing invalid input

emilyy-dev opened the pull request:
`DurationParser` using that regex and `Matcher#find()` only finds the pattern in the given input string, while the pattern is right for individual segments denoting value-unit sequences, it does not prevent the input to have non-matching content.
This patch fixes that by manually parsing the input.

Additionally it conforms to the recommended way of parsing an argument: peek input, pop input on successful parsing.

emilyy-dev committed:
Fix `DurationParser` allowing invalid input by not using regex and manually parsing the input

emilyy-dev committed:
Do not parse time amount as unsigned long

jpenilla commented:
> Additionally it conforms to the recommended way of parsing an argument: peek input, pop input on successful parsing.

This isn't the case in 2.x, the tree will restore the cursor on a failed parse. You only need to do this if your parsing logic calls for it.

emilyy-dev commented:
> > Additionally it conforms to the recommended way of parsing an argument: peek input, pop input on successful parsing.
> 
> This isn't the case in 2.x, the tree will restore the cursor on a failed parse. You only need to do this if your parsing logic calls for it.

In that case the docs need updating 👍🏻 I'll revert that change here for now

emilyy-dev committed:
Address comment

emilyy-dev committed:
Readd parsed format to class javadoc

Citymonstret commented:
> > > Additionally it conforms to the recommended way of parsing an argument: peek input, pop input on successful parsing.
> > 
> > 
> > This isn't the case in 2.x, the tree will restore the cursor on a failed parse. You only need to do this if your parsing logic calls for it.
> 
> In that case the docs need updating 👍🏻 I'll revert that change here for now

Yes, the docs should be updated. When we switched over to the `CommandInput` there wasn't exactly a plan to no longer require peeking. That was introduced later on to increase the comfort of using the new system. I haven't done a thorough sweep of the documentation to reflect this (although it is at the very least pointed out in the external docs now)

emilyy-dev committed:
Handle case where input duration overflows

Citymonstret committed:
Merge branch 'master' into fix/duration-parser-allows-invalid-input

jpenilla committed:
fix: restrict duration parsing to ASCII digits

The previous regex only accepted ASCII digits, while Character.isDigit
also recognizes Unicode decimal digits. Avoid accidentally expanding the
accepted duration grammar, and add a regression test for full-width
digits.

jpenilla committed:
Merge branch 'master' into fix/duration-parser-allows-invalid-input
